File: Count3s.cpp
// Name: Tony Tam
// Course: CS177 Software Engineering
// CRN: 75874
// Date: 09/11/2020
// Assignment: Count3s.cpp

//Include libraries
#include <iostream>
#include <string>
#include <fstream>
#include <cstdint>

//using namespace std, reduce code verbage
using namespace std;

int main() {

	//initialize counter starting at 0
	int counter = 0;

	//setup empty unintiallize uint32_t variable from cstdint library
	uint32_t theNumber;

	//setup input stream using ifstream variable with name of file in first parameter and input modes in the second
	ifstream inputStream("threesData.bin", ios::in | ios::binary);

	//handle file opening errors
	if (!inputStream.is_open()) {
		cout << "Cannot Open File";
	}
	else {

		inputStream.seekg(0, ios::end); //seekg() sets fileposition from the EndOfFile direction to Begining at position 0
		int sizeOfFile = (int)inputStream.tellg(); //cast filesize position as integer
		inputStream.seekg(0, ios::beg); //reset streampos to the beginning
		
		int numberOfNumbers = 0;

		while (inputStream.tellg() < sizeOfFile) { //tellg returns current file position
			inputStream.read((char*) &theNumber, sizeof(theNumber)); //read size of uint32_t into "theNumber"
			//cout << theNumber << endl;
			if (theNumber == 3) { //if theNumber is equal to 3 increment the counter by 1
				//cin;
				counter++;
				//cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
				//cin.clear();
			}

			numberOfNumbers++;
		}

		cout << numberOfNumbers << endl;
		cout << counter << endl; //print out counter

		inputStream.close(); //close input stream
	}

}

File: Count3s.py
# Name: Tony Tam
# Course: CS177 Software Engineering
# CRN: 75874
# Date: 09/11/2020
# Assignment: Count3s.py

import struct

#open the file with file name in read and binary mode
inputStream = open("threesData.bin", "rb")
inputStream.seek(0,2) #set file pointer to end of file
sizeOfFile = inputStream.tell() # get size of file
inputStream.seek(0,0) #reset file pointer to begin

counter = 0 #set counter to 0

while inputStream.tell() < sizeOfFile:

    data = inputStream.read(4) #read 4 Bytes into data
        
    integer = struct.unpack('I', data) #using stuct unpack function to convert data to tuple containing an integer.
    
    if int(integer[0]) == 3: # access tuple position 0 and compare to 3
        counter += 1 # increment counter

        
print(counter) # print counter